tests/ai/test_utils.py

```diff
@@ -35,3 +51,379 @@ def test_column_type_to_python_type_errors():
+                        "type_json": '{"name":"d","type":"interval day to second","nullable":true,"metadata":{}}',
+                        "type_name": "INTERVAL",
+                        "type_precision": 0,
+                        "type_scale": 0,
+                        "position": 3,
+                        "parameter_type": "PARAM",
+                    }
+                ]
+            ),
+            [{"d": datetime.timedelta(days=1, seconds=1, microseconds=123456)}, {"d": None}],
+        ),
+        (
+            generate_function_info(
+                [
+                    {
+                        "name": "e",
+                        "type_text": "timestamp",
+                        "type_json": '{"name":"e","type":"timestamp","nullable":true,"metadata":{}}',
+                        "type_name": "TIMESTAMP",
+                        "type_precision": 0,
+                        "type_scale": 0,
+                        "position": 4,
+                        "parameter_type": "PARAM",
+                    }
+                ]
+            ),
+            [{"e": datetime.datetime(2024, 9, 11)}, {"e": "2024-09-11T00:00:00"}, {"e": None}],
+        ),
+        (
+            generate_function_info(
+                [
+                    {
+                        "name": "f",
+                        "type_text": "timestamp_ntz",
+                        "type_json": '{"name":"f","type":"timestamp_ntz","nullable":true,"metadata":{}}',
+                        "type_name": "TIMESTAMP_NTZ",
+                        "type_precision": 0,
+                        "type_scale": 0,
+                        "position": 5,
+                        "parameter_type": "PARAM",
+                    }
+                ]
+            ),
+            [{"f": datetime.datetime(2024, 9, 11)}, {"f": "2024-09-11T00:00:00"}, {"f": None}],
+        ),
+        (
+            generate_function_info(
+                [
+                    {
+                        "name": "g",
+                        "type_text": "date",
+                        "type_json": '{"name":"g","type":"date","nullable":true,"metadata":{}}',
+                        "type_name": "DATE",
+                        "type_precision": 0,
+                        "type_scale": 0,
+                        "position": 6,
+                        "parameter_type": "PARAM",
+                    }
+                ]
+            ),
+            [{"g": datetime.date(2024, 9, 11)}, {"g": "2024-09-11"}, {"g": None}],
+        ),
+        (
+            generate_function_info(
+                [
+                    {
+                        "name": "h",
+                        "type_text": "map<string,array<int>>",
+                        "type_json": '{"name":"h","type":{"type":"map","keyType":"string","valueType":{"type":"array","elementType":"integer","containsNull":true},"valueContainsNull":true},"nullable":true,"metadata":{}}',
```

Comment:
can we make this json for readability?

Reply:
Actually I'm copying over from the function info got back from a get_function request, honestly I don't know what required fields are as there's no single documentation, so I don't think we can construct this field ourselves, I'd like to keep it as a string here